File: HW4/not_lasso.R
n = 50
p = 50
s = 10
T = 3000
X = matrix(rnorm(n*p),nrow = n)
beta_true = matrix(rep(0,p), nrow = p)
beta_true[1:s] = 1:s
Y=X%*%beta_true + rnorm(n)
db = rep(0,p)
beta = matrix(rep(0,p), nrow = p)
beta_all = matrix(rep(0, p*T), nrow = p)
epsilon = 0.0001
R = Y
for(t in 1:T)
{
  for(j in 1:p)
  {
    db[j] = sum(R*X[,j])
  }
j = which.max(abs(db))
beta[j] = beta[j] + db[j]*epsilon
R = R - X[,j]*db[j]*epsilon
beta_all[, t] = beta
}
matplot(t(matrix(rep(1, p), nrow = 1)%*%abs(beta_all)), t(beta_all), type = 'l')

File: HW4/not_lasso.py
import numpy as np
import matplotlib.pyplot as plt

n = 50
p = 50
s = 10
T = 3000
epsilon = 0.0001
beta_all = np.zeros((p,T))
X = np.random.standard_normal((n,p))
beta_true = np.zeros(p)
beta_true[0:s] = range(1, s+1)
Y = np.add(np.dot(X,beta_true),np.random.standard_normal(n,))
db = np.zeros(p)
beta = np.zeros(p)
R = Y
for t in range(0,T):
    for j in range(0,p):
        db[j] = np.sum(R*X[:, j])
    j = np.argmax(db)
    beta[j] = beta[j] + db[j]*epsilon
    R = R - X[:,j]*db[j]*epsilon
    for i in range(0,p):
        beta_all[i][t] = beta[i]

u = np.transpose(np.dot(np.ones((1,p)),abs(beta_all)))
v = np.transpose(beta_all)
plt.figure()
plt.plot(u, v,  label='Stagewise Regression')